### PR TITLE
Fix: System exports() should wrap expressions with pure annotation

### PR DESCRIPTION
**Pull request by TrySound**
# Update

by @lukastaegert
This PR fixes the original issue reported below by moving any trailing comments and whitespace inside the `exports` function when rendering named export declarations or assignments to exports. For default exports, everything was already working correctly but I added a test nonetheless.

Other changes:
* cleaned up system binding rendering by moving the code for rendering assignments and update expressions from `Identifier` to the corresponding AST nodes.
* made sure the `_actual` files and folders are always removed before running a test, which fixes an issue with chunking form tests where the new output was not necessarily overwriting the old output due to the hashing

# Original PR
This took me a day to figure out why parts of my vendors
are just not exported across chunks.

Take this module
https://unpkg.com/react-head@2.2.0/dist/index.esm.js

`HeadCollector` will be wrapped with `exports` this way
so uglify will consider exports is pure function and will
happily remove it as unused.

```js
var HeadCollector =
/*#__PURE__*/
exports('id', function (_Component) {
  _inheritsLoose(HeadCollector, _Component);
...
```

Note: this is just a test

### CODE REVIEW

**Comment by TrySound** [comment]
This is quite critical bug for apps where code splitting is usually used.

**Comment by guybedford** [comment]
Where is the System chunk output in this test? Or is that the bug?

File: test/chunking-form/samples/system-exports-wrap-pure-annotation/_expected/system/chunk-85f39e10.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+System.register([], function (exports, module) {
+	'use strict';
+	return {
+		execute: function () {
+
+			var dep =
+				/*#__PURE__*/
+				exports('a', (function() {
```
Comment (TrySound) [inline test/chunking-form/samples/system-exports-wrap-pure-annotation/_expected/system/chunk-85f39e10.js]: it's here

**Comment by guybedford** [comment]
This is a tough one - basically any transformation in any build tool is vulnerable to this `/#__PURE__*/` comment, so a specific fix for system feels a bit too specific.

I really dislike this whole `/*#__PURE__*/` thing, and wish it did not exist personally.... can it not be disabled in uglify options?

**Comment by TrySound** [comment]
Sadly disabling it doesn't work for some reason. It helps with transpiled classes and I do not see any other format is affected.

I think the best solution will be moving exports to the end of file like we do with all other formats.
```js
var dep =
	/*#__PURE__*/
	(function() {
		return 0;
	})();
 exports('aa', dep);
```

**Comment by guybedford** [comment]
@TrySound unfortunately that is not an option as the whole point of the System format is getting the live exports binding support. Do you know why uglify isn't disabling the PURE option... may be worth looking into that?

**Comment by TrySound** [comment]
Not yet, will investigate.

And still. Rollup isn't able to treeshake transpiled classes so I rely on uglify here. This trick is used by [size-snapshot](https://github.com/TrySound/rollup-plugin-size-snapshot) to prove treeshakability (dead code eliminating). Without this feature our apps with blow up with large size.

**Comment by lukastaegert** [comment]
Semantically, it would actually make sense to move the comments preceding exports into the wrapper code. E.g. for tree-shaking, we already have an elaborate logic in place that makes sure, preceding and trailing comments are removed together with statements.

This is handled via `NodeRenderOptions` that pass a `start`/`end` to the child. This logic could be extended so that e.g. `AssignmentExpression` also passes `start`/`end` to the child. `Identifier` could then use these to render the correct boundaries and include the comment inside the wrapper.

**Comment by guybedford** [comment]
@lukastaegert that sounds like it could work.

**Comment by lukastaegert** [comment]
Added a fix to the PR

**Comment by lukastaegert** [comment]
@TrySound please give this a try if it fixes your issue.

**Comment by guybedford** [comment]
There seems to be an odd failure in Node 6 here - https://circleci.com/gh/rollup/rollup/359?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link. Which doesn't seem like a trivial CI issue, and maybe a Node 6 compatibility issue? Seems strange that it would fail in this way.

**Comment by guybedford** [comment]
Ah it seems the failure is on master, tracking it down now.

**Comment by guybedford** [comment]
Merging anyway, as can confirm this is building in Node 6 fine.

**Pull request merged**